File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionFadeIn.java
package br.com.insanitech.spritekit;

/**
 * Created by anderson on 06/01/17.
 */

class SKActionFadeIn extends SKAction {
    private float startAlpha;

    @Override
    void computeStart() {
        startAlpha = getParent().alpha;
    }

    @Override
    void computeAction(long elapsed) {
        switch (timingMode) {
            case Linear:
                getParent().alpha = SKEaseCalculations.linear(elapsed, startAlpha, 1.0f - startAlpha, duration);
                break;
            case EaseIn:
                getParent().alpha = SKEaseCalculations.easeIn(elapsed, startAlpha, 1.0f - startAlpha, duration);
                break;
            case EaseOut:
                getParent().alpha = SKEaseCalculations.easeOut(elapsed, startAlpha, 1.0f - startAlpha, duration);
                break;
            case EaseInEaseOut:
                getParent().alpha = SKEaseCalculations.easeInOut(elapsed, startAlpha, 1.0f - startAlpha, duration);
                break;
        }
    }

    @Override
    void computeFinish() {
        getParent().alpha = 1.0f;
    }

    @Override
    boolean willHandleFinish() {
        return false;
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionFadeIn.kt
package br.com.insanitech.spritekit

/**
 * Created by anderson on 06/01/17.
 */

internal class SKActionFadeIn : SKAction() {
    private var startAlpha: Float = 0.toFloat()

    internal override fun computeStart() {
        startAlpha = parent?.alpha ?: 0.0f
    }

    internal override fun computeAction(elapsed: Long) {
        when (timingMode) {
            SKActionTimingMode.Linear -> parent?.alpha = SKEaseCalculations.linear(elapsed.toFloat(), startAlpha, 1.0f - startAlpha, duration.toFloat())
            SKActionTimingMode.EaseIn -> parent?.alpha = SKEaseCalculations.easeIn(elapsed.toFloat(), startAlpha, 1.0f - startAlpha, duration.toFloat())
            SKActionTimingMode.EaseOut -> parent?.alpha = SKEaseCalculations.easeOut(elapsed.toFloat(), startAlpha, 1.0f - startAlpha, duration.toFloat())
            SKActionTimingMode.EaseInEaseOut -> parent?.alpha = SKEaseCalculations.easeInOut(elapsed.toFloat(), startAlpha, 1.0f - startAlpha, duration.toFloat())
        }
    }

    internal override fun computeFinish() {
        parent?.alpha = 1.0f
    }

    internal override fun willHandleFinish(): Boolean {
        return false
    }
}
